### ISSUE TITLE
First example in docs doesn't work (or I'm understanding it wrong?)

### ISSUE THREAD

**Issue by jdanbrown** [pex-tool/pex#157]
https://pex.readthedocs.org/en/stable/buildingpex.html says:

> The easiest way to build .pex files is with the pex utility, which is made available when you pip install pex. Do this within a virtualenv, then you can use pex to bootstrap itself:
> 
> ```
> $ pex pex requests -c pex -o ~/bin/pex
> ```
> 
> This command creates a pex file containing pex and requests, using the console script named “pex”, saving it in ~/bin/pex. At this point, assuming ~/bin is on your $PATH, then you can use pex in or outside of any virtualenv.

But when I try building a .pex that way, it isn't able to `import requests`:

```
$ pex pex requests -c pex -o /tmp/pex
$ /tmp/pex
Python 2.7.9 (default, Feb 10 2015, 03:29:19)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named requests
```

Using pex-1.0.3:

```
$ pip show pex

---
Metadata-Version: 2.0
Name: pex
Version: 1.0.3
Summary: The PEX packaging toolchain.
Home-page: https://github.com/pantsbuild/pex
Author: UNKNOWN
Author-email: UNKNOWN
License: Apache License, Version 2.0
Location: /usr/local/lib/python2.7/site-packages
Requires: setuptools
```

**Comment by kwlzn** [comment]
@jdanbrown this is because the default behavior of `pex` is to drop you into a repl. in this case, you're successfully building the pex tool (as a pex), but in the encapsulated pex you've built you're invoking it with an empty environment and as a repl (`/tmp/pex`). if you were to do `/tmp/pex requests` here you would be able to import `requests`:

```
$ pex --version
pex 1.0.3

$ pex
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named requests
>>> ^D

$ pex pex requests -c pex -o /tmp/pex.pex

$ /tmp/pex.pex --version
pex.pex 1.0.3

$ /tmp/pex.pex
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named requests
>>> ^D

$ /tmp/pex.pex requests
import Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
>>> ^D
```

**Comment by jsirois** [comment]
Some bug to do with -c:
```
$ pex pex requests -o /tmp/pex
$ /tmp/pex
Python 3.4.3 (default, Mar 25 2015, 17:13:50) 
[GCC 4.9.2 20150304 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
>>> 
```

**Comment by jsirois** [comment]
..or not:
```
$ pex pex requests -c pex -o /tmp/pex
$ PEX_INTERPRETER=1 /tmp/pex
Python 3.4.3 (default, Mar 25 2015, 17:13:50) 
[GCC 4.9.2 20150304 (prerelease)] on linux
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
>>> 
```

**Comment by jsirois** [comment]
@jdanbrown are you up for more abuse or do you want one of use to figure it out from there?

**Comment by kwlzn** [comment]
to further clarify: this might be a little confusing because the intent of 'using pex to bootstrap itself' is to produce a pex'd pex cli that acts exactly like the pex cli that is produced on the $PATH when you `pip install pex`, except that it's self-contained (i.e. you could now distribute this pex.pex without having to pip install anything - incl dependencies of pex - on other machines and furthermore could now `pip uninstall pex`). this is completely optional and you'd really only ever do this once.

once you do this tho, invoking the resulting pex'd pex cli (`/tmp/pex.pex`) should behave exactly like the 'pip install pex' version (`pex`). hopefully this is made clear in the example above, but let me know if you have other questions.

and FWIW, if all you're really trying to do here is get requests into a pex, then you'd just do:

```
$ pex requests -o /tmp/requests.pex
$ /tmp/requests.pex
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
>>> ^D
```

**Comment by jsirois** [comment]
I think though that `pex pex [any other requirements whatsoever] -c pex -o pex.pex` should produce a pex.pex with an entrypoint that is pex's and thus running `./pex.pex` would never land you in an interpreter, but instead just display pex help. That's the bug here.

**Comment by jsirois** [comment]
Actually, the pex script installed in a venv with pex run with with no arguments drops you into an interpeter.... which is odd, but consistent with the behavior of the pex.pex.  So it's that bit that is both odd, and still maybe broken wrt not making requests available.  Its certainly in the pex.pex .deps/ dir.
```
$ unzip -qc /tmp/pex PEX-INFO | python -mjson.tool
{
    "always_write_cache": false,
    "build_properties": {
        "class": "CPython",
        "platform": "linux-x86_64",
        "version": [
            3,
            4,
            3
        ]
    },
    "code_hash": "da39a3ee5e6b4b0d3255bfef95601890afd80709",
    "distributions": {
        "pex-1.0.3-py2.py3-none-any.whl": "9b479c2967fcb9866cb8b019740e265c9bc00da2",
        "requests-2.7.0-py2.py3-none-any.whl": "29d2c9ec33391aeb79f237ea5916c42d97b45a7b",
        "setuptools-15.2-py2.py3-none-any.whl": "b0059e5095f9e683017bcc691f9cae26c19c3796"
    },
    "entry_point": "pex.bin.pex:main",
    "ignore_errors": false,
    "inherit_path": false,
    "requirements": [
        "pex==1.0.3",
        "setuptools==15.2",
        "requests==2.7.0"
    ],
    "zip_safe": true
}
$ zipinfo -1 /tmp/pex | grep .deps/ | grep requests | head
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/DESCRIPTION.rst
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/METADATA
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/RECORD
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/WHEEL
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/metadata.json
.deps/requests-2.7.0-py2.py3-none-any.whl/requests-2.7.0.dist-info/top_level.txt
.deps/requests-2.7.0-py2.py3-none-any.whl/requests/__init__.py
.deps/requests-2.7.0-py2.py3-none-any.whl/requests/adapters.py
.deps/requests-2.7.0-py2.py3-none-any.whl/requests/api.py
.deps/requests-2.7.0-py2.py3-none-any.whl/requests/auth.py
```

So there is some oddness indeed.

**Comment by kwlzn** [comment]
dropping into a repl has been the default for `pex` when not passed `-o` for as long as I can remember:

```
$ pex pex==0.7.0 -c pex -o /tmp/pex
$ /tmp/pex
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> ^D
```

and I believe that it's the normal pex environment isolation you're seeing here that is inherently not providing for access to requests when the repl entrypoint is invoked:

```
$ sudo pip2.7 install requests --upgrade
Requirement already up-to-date: requests in /Library/Python/2.7/site-packages
Cleaning up...
$ python2.7 -c 'import requests; print requests.__version__'
2.7.0
$ pex
Python 2.7.10 (default, Jul 14 2015, 19:46:27) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.39)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import requests
Traceback (most recent call last):
  File "<console>", line 1, in <module>
ImportError: No module named requests
>>> ^D
```

so afaict the behavior seen here is as expected. the execution of the pex console script passed via `-c` (now the main entrypoint for pex.pex) is just dropping you into an _isolated_ repl because you've provided it with no dependency inputs for the new execution context.

**Comment by jsirois** [comment]
That isolation does not jive with PEX-INFO though.

**Comment by kwlzn** [comment]
@jsirois presumably due to the nature of pex itself the console script run is building and execing in a net-new (and thus isolated) pex context - e.g. if you did `/tmp/pex.pex requests -o /tmp/requests.pex` the result would be to write a net-new PEX-INFO to /tmp/requests.pex. afaict, thats the equivalent of whats happening under the covers in the repl case without the `-o`.

I do think I see what you're getting at re: oddness of the repl - in that there's no real point to running 'pex' and getting a bare repl - other than to fundamentally know that pex is working (and perhaps a shortcut to just running a scrubbed down python interpreter). dunno, I kinda like the bare repl behavior as a baseline if nothing else.

**Comment by jsirois** [comment]
To sharpen this a bit for my own mind, for another sdist with a script - instead of pex, pants:
```
$ mkdir /tmp/pants && touch /tmp/pants/pants.ini && cd /tmp/pants
$ pyenv activate pex
(pex) $ pex --version
pex 1.0.3
(pex) $ pex pantsbuild.pants fabric -c pants -o pants.pex --python=python2.7
(pex) $ pyenv deactivate 
$ ./pants.pex 
No goals specified.
Use `pants goals` to list goals.
Use `pants help` to get help.
```

And:
```
$ PEX_INTERPRETER=1 ./pants.pex 
Python 2.7.10 (default, May 26 2015, 04:16:29) 
[GCC 5.1.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import fabric
>>> 
```

So the weird thing is 2-fold.  The 1st bit is `pex -h` or `pex [any args]` runs the pex tool, but `pex` does something different.  And the 2nd bit is that the `pex` drops you to an interpreter that is not the same as `PEX_INTERPRETER=1 pex`, that's different than every other pex, ie:
```
(pex) $ pex pantsbuild.pants fabric -o pants.pex --python=python2.7 --no-wheel
(pex) $ pyenv deactivate 
$ ./pants.pex 
Python 2.7.10 (default, May 26 2015, 04:16:29) 
[GCC 5.1.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import fabric
>>> 
$ PEX_INTERPRETER=1 ./pants.pex 
Python 2.7.10 (default, May 26 2015, 04:16:29) 
[GCC 5.1.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import fabric
>>> 
```

**Comment by jsirois** [comment]
Ah - crossed streams a bit.  I think we agree the behavior of the pex entrypoint is the odd duck here, like it or no.

**Comment by jsirois** [comment]
Ah - and yes - what you say makes logic if not sense.  `pex` does not in fact drop you into a repl like `PEX_INTERPRETER=1 pex` does, it creates a whole new pex quickly enough not to be noticeable and drops you into that ephemeral pex's PEX_INTERPRETER=1.

**Comment by wickman** [comment]
Yes, this is working as intended.  But what's intended is surprising when '-c pex' is used.  It's a bad example.

That being said, I see that there's been vigorous debate on the behavior of 'pex' without args.  We could add a --repl flag that forces the repl (and is incompatible with any entry point options like -c/-m) and change the default behavior to show help if no requirements are specified.  This at least allows you to do what pex can do for you now: drop you into a vanilla python interpreter if it can find the one specified (e.g. "pex --python=python3.2" => "pex --python=python3.2 --repl".)

I like this less from a logical purity point of view but logical purity is rarely the best user experience.

**Comment by jsirois** [comment]
One way to mitigate user surprise, maybe, is to customize pex repls.  If you had a custom banner that included the pex env path and maybe also the list of installed requirements - then at least when I run `./pex.pex` I'd see:
```console
Welcome to your pex environment!
container: /tmp/abc123
interpreter: python2.7
environment: []

>>>
```

And maybe I'd notice /tmp/abc123 != ./pex.pex.

That said, this problem would go away completely for all but the very interested if pex distributed a pex.pex.  Then no one who might be surprised by the no-arg behavior would need to ever produce the pex that exhibited it.  They'd just download pex.pex and not expect it to have requests, they'd just expect the CLI.

**Comment by thalin** [comment]
So this just tripped me up; I didn't realize that pex w/o args drops to an interpreter. As a dumb user, I expected some help output or something telling me I was doing something dumb. Instead, I was expected to be smart enough to realize that pex by itself acts like running the python it's using.

**Comment by jsirois** [comment]
It's about time to implement something like the REPL customization described here: https://github.com/pex-tool/pex/issues/157#issuecomment-138969369

Although I'm long past being confused by the behavior of `pex` - it is definitely still objectively confusing from any reasonably newcomerish stance.

**Comment by jsirois** [comment]
For a PEX build like so:
:; python -mpex --include-tools requests -orequests.pex --layout packed


Running `python requests.pex` gives this currently:
```pycon
Python 3.11.9 (main, Apr 26 2024, 19:20:24) [GCC 13.2.0] on linux
Pex 2.14.0 hermetic environment with 1 requirement and 5 activated distributions.
Type "help", "pex_info", "copyright", "credits" or "license" for more information.
>>> pex_info
Type pex_info() for information about this PEX, or pex_info(json=True) for even more details.
>>> pex_info()
Running from packed PEX directory: ./requests.pex
Requirements:
  requests
Activated Distributions:
  /home/jsirois/.pex/installed_wheels/70761cfe03c773ceb22aa2f671b4757976145175cdfca038c02654d061d6dcc6/requests-2.32.3-py3-none-any.whl
  /home/jsirois/.pex/installed_wheels/753f10e867343b4511128c6ed8c82f7bec3bd026875576dfd88483c5c73b2fd8/charset_normalizer-3.3.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
  /home/jsirois/.pex/installed_wheels/82fee1fc78add43492d3a1898bfa6d8a904cc97d8427f683ed8e798d07761aa0/idna-3.7-py3-none-any.whl
  /home/jsirois/.pex/installed_wheels/a448b2f64d686155468037e1ace9f2d2199776e17f0a46610480d311f73e3472/urllib3-2.2.2-py3-none-any.whl
  /home/jsirois/.pex/installed_wheels/c198e21b1289c2ab85ee4e67bb4b4ef3ead0892059901a8d5b622f24a1101e90/certifi-2024.7.4-py3-none-any.whl
This PEX includes tools. Exit the repl (type quit()) and run `PEX_TOOLS=1 ./requests.pex/__main__.py -h` for tools help.
>>>
```

Or, in the OP case, just running Pex with no args (`python -mpex` in this local hack):
```pycon
Python 3.11.9 (main, Apr 26 2024, 19:20:24) [GCC 13.2.0] on linux
Pex 2.14.0 hermetic environment with no dependencies.
Type "help", "pex_info", "copyright", "credits" or "license" for more information.
>>> pex_info()
Running from loose PEX directory: /tmp/tmpkbf3ssi3
>>>
```

**Comment by jsirois** [comment]
Actually, its easy enough to know when the PEX is the Pex main running an ephemeral no-args PEX yielding:
```pycon
:; pex
Python 3.11.9 (main, Apr 26 2024, 19:20:24) [GCC 13.2.0] on linux
Pex 2.14.1 hermetic environment with no dependencies. Run `pex -h` for Pex CLI help.
Type "help", "pex_info", "copyright", "credits" or "license" for more information.
```

Here the `Run `pex -h` for Pex CLI help.` suffix on line 2 of the banner is the new affordance in this special case. It may make sense to later follow up with colors, when appropriate, to help highlight that this is not a normal Python REPL, but this seems like a good step forward as-is.

I still need to work out the `--venv` case, which has its own entrypoint and will probably get slightly different / simpler REPL affordances.

**Comment by jsirois** [comment]
Ok. So, for the OP, you now get:
![image](https://github.com/user-attachments/assets/47187fc5-86b9-403f-940f-d002baaece88)
Note the "Run `dist/pex -h` for Pex CLI help." affordance.

And for a PEX with content:
```
:; pex --interpreter-constraint ">=3.8" --pip-version latest --sh-boot requests -orequests.pex
```
![image](https://github.com/user-attachments/assets/4ba83eac-9ed0-45a3-a681-28f4da9d1ed4)

**Comment by jsirois** [comment]
Alright @jdanbrown, for your OP example, but with the just released Pex 2.15.0:
```
:; pex -V
2.15.0
:; pex pex requests -c pex -o /tmp/pex
```

You now see this:
![image](https://github.com/user-attachments/assets/a111bb89-a8fd-457a-a7c0-3695249e5892)

I'm not sure if that would have helped avoid the confusion, but hopefully its better.
